storage/memory/peer_store.go

```diff
@@ -484,26 +489,61 @@ func (ps *peerStore) AnnouncePeers(ih bittorrent.InfoHash, seeder bool, numWant
 	return
 }
 
-func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+func (ps *peerStore) fullscrapeSwarms(addressFamily bittorrent.AddressFamily) (resp []bittorrent.Scrape) {
+	shards := ps.shards[:len(ps.shards)/2]
+	if addressFamily == bittorrent.IPv6 {
+		shards = ps.shards[len(ps.shards)/2:]
+	}
+
+	for _, shard := range shards {
+		shard.RLock()
+		for ih, swarm := range shard.swarms {
+			resp = append(resp, bittorrent.Scrape{
+				InfoHash:   ih,
+				Complete:   uint32(len(swarm.seeders)),
+				Incomplete: uint32(len(swarm.leechers)),
+			})
+		}
+		shard.RUnlock()
+		// Not sure if this helps...
+		runtime.Gosched()
+	}
+
+	return
+}
+
+func (ps *peerStore) ScrapeSwarms(infoHashes []bittorrent.InfoHash, addressFamily bittorrent.AddressFamily) (resp []bittorrent.Scrape) {
 	select {
 	case <-ps.closed:
 		panic("attempted to interact with stopped memory store")
 	default:
 	}
 
-	resp.InfoHash = ih
-	shard := ps.shards[ps.shardIndex(ih, addressFamily)]
-	shard.RLock()
-
-	swarm, ok := shard.swarms[ih]
-	if !ok {
-		shard.RUnlock()
+	if len(infoHashes) == 0 {
+		// This is a fullscrape.
+		start := time.Now()
+		resp = ps.fullscrapeSwarms(addressFamily)
+		recordFullscrapeDuration(time.Since(start))
```

Comment:
Couldn't we pull the metric timing up to the caller of this function? Are we doing that for the non-full-scrape case? The caller can definitely make the check of whether the `len(infoHashes) == 0`

Reply:
Well, we're tracking some request durations in the frontends, if enabled via the config flag.

I kinda see your point, as in: The prometheus metrics for the storage are all storage-internal, as in not visible to a frontend. I just added it here because I thought it would be interesting to see, similar to GC duration.

I expect fullscrapes to be an occasional thing, and I know from some people who didn't enable the request timing in the frontend (because it costs performance).

Opinion?